src/main/java/co/cask/gcp/datastore/source/util/DatastoreSourceQueryUtil.java

```diff
@@ -0,0 +1,368 @@
+   */
+  public static String transformKeyToKeyString(Key key, SourceKeyType keyType) {
+    switch (keyType) {
+      case KEY_LITERAL:
+        StringBuilder builder = new StringBuilder("key(");
+
+        key.getAncestors()
+          .forEach(a -> appendKindWithNameOrId(builder, a.getKind(), a.getNameOrId()).append(", "));
+
+        appendKindWithNameOrId(builder, key.getKind(), key.getNameOrId());
+        return builder.append(")").toString();
+
+      case URL_SAFE_KEY:
+        return key.toUrlSafe();
+
+      default:
+        throw new IllegalArgumentException(
+          String.format("Unable to transform key [%s] to type [%s] string representation", key, keyType.getValue()));
+    }
+  }
+
+  /**
+   * Constructs protobuf query instance which will be used for query splitting.
+   * Adds ancestor and property filters if present in the given configuration.
+   *
+   * @param config Datastore configuration
+   * @return protobuf query instance
+   */
+  public static com.google.datastore.v1.Query constructPbQuery(DatastoreSourceConfig config) {
+    com.google.datastore.v1.Query.Builder builder = com.google.datastore.v1.Query.newBuilder()
+      .addKind(KindExpression.newBuilder()
+                 .setName(config.getKind()));
+
+    List<Filter> filters = config.getFilters().entrySet().stream()
+      .map(e -> DatastoreHelper.makeFilter(e.getKey(), PropertyFilter.Operator.EQUAL,
+            constructFilterValue(e.getKey(), e.getValue(), config.getSchema()))
+        .build())
+      .collect(Collectors.toList());
+
+    List<PathElement> ancestors = config.getAncestor();
+    if (!ancestors.isEmpty()) {
+      filters.add(DatastoreHelper.makeAncestorFilter(constructKey(ancestors, config)).build());
+    }
+
+    if (!filters.isEmpty()) {
+      builder.setFilter(DatastoreHelper.makeAndFilter(filters));
+    }
+
+    return builder.build();
+  }
+
+  /**
+   * Transforms protobuf query instance emitted by the query splitter into {@link Query<Entity>}.
+   *
+   * @param pbQuery protobuf query instance
+   * @param config Hadoop configuration
+   * @return query instance
+   */
+  public static Query<Entity> transformPbQuery(com.google.datastore.v1.Query pbQuery, Configuration config) {
+    EntityQuery.Builder builder = Query.newEntityQueryBuilder()
+      .setNamespace(config.get(DatastoreSourceConstants.CONFIG_NAMESPACE))
+      .setKind(config.get(DatastoreSourceConstants.CONFIG_KIND));
+
+    if (pbQuery.hasFilter()) {
+      builder.setFilter(transformFilterFromPb(pbQuery.getFilter()));
+    }
+
+    return builder.build();
+  }
+
+  /**
+   * Constructs Datastore protobuf key instance based on given list of path elements
+   * and Datastore configuration.
+   *
+   * @param pathElements list of path elements
+   * @param config Datastore configuration
+   * @return Datastore protobuf key instance
+   */
+  private static com.google.datastore.v1.Key constructKey(List<PathElement> pathElements,
```

Comment:
unless these methods are called from more than one non-test class, I think it makes more sense to move them directly to either the config class, or whatever class is calling them.